jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java

```diff
@@ -0,0 +1,238 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.gradle;
+
+import com.google.api.client.http.HttpTransport;
+import com.google.cloud.tools.jib.builder.BuildConfiguration;
+import com.google.cloud.tools.jib.builder.BuildImageSteps;
+import com.google.cloud.tools.jib.builder.SourceFilesConfiguration;
+import com.google.cloud.tools.jib.cache.Caches;
+import com.google.cloud.tools.jib.image.ImageReference;
+import com.google.cloud.tools.jib.image.InvalidImageReferenceException;
+import com.google.cloud.tools.jib.image.json.BuildableManifestTemplate;
+import com.google.cloud.tools.jib.registry.RegistryClient;
+import com.google.common.base.Preconditions;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.GradleException;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
+import org.gradle.api.tasks.TaskAction;
+
+/** Builds a container image. */
+public class BuildImageTask extends DefaultTask {
+
+  /** Directory name for the cache. The directory will be relative to the build output directory. */
+  private static final String CACHE_DIRECTORY_NAME = "jib-cache";
+
+  /** {@code User-Agent} header suffix to send to the registry. */
+  private static final String USER_AGENT_SUFFIX = "jib-gradle-plugin";
+
+  @Nullable private ImageConfiguration from;
+  @Nullable private ImageConfiguration to;
+  @Nullable private List<String> jvmFlags;
+  @Nullable private String mainClass;
+  private boolean reproducible;
+  @Nullable private Class<? extends BuildableManifestTemplate> format;
+  private boolean useOnlyProjectCache;
+
+  @Nested
+  @Nullable
+  public ImageConfiguration getFrom() {
+    return from;
+  }
+
+  @Nested
+  @Nullable
+  public ImageConfiguration getTo() {
+    return to;
+  }
+
+  @Input
+  @Nullable
+  public List<String> getJvmFlags() {
+    return jvmFlags;
+  }
+
+  @Input
+  @Nullable
+  @Optional
+  public String getMainClass() {
+    return mainClass;
+  }
+
+  @Input
+  public boolean getReproducible() {
+    return reproducible;
+  }
+
+  @Input
+  @Nullable
+  public Class<? extends BuildableManifestTemplate> getFormat() {
+    return format;
+  }
+
+  @Input
+  public boolean getUseOnlyProjectCache() {
+    return useOnlyProjectCache;
+  }
+
+  @TaskAction
+  public void buildImage() throws InvalidImageReferenceException, IOException {
+    // Asserts required inputs are not null.
```

Comment:
This comment is a bit confusing ("required inputs" sounds like it means "required config params", even though the params it's checking aren't required). Maybe change this to be more along the lines of "Asserts defaults have been set on configuration parameters".

Reply:
Oh wait, this was actually referring to the `@Input`-annotated closure-configurable parameters. Here, it asserts that the required `@Input`s (the ones that don't have `@Optional` and are non-primitives) are not `null`. That's why `mainClass` is not in this list.

Reply:
Also, this PR is after #180 so I'll make a change there.

Reply:
Scratch that, this change is actually introduced in this PR haha.